### PR TITLE
ZAP API generator for Ruby

### PR DESCRIPTION
**Pull request by tjackiw**
Here's a Ruby generator for the ZAP API.

The generated files will be part of the `zap-api-ruby` repo and will be published as an official gem on [https://rubygems.org/](https://rubygems.org/). Thanks!

### CODE REVIEW

File: src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
+package org.zaproxy.zap.extension.api;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java]: Could you add the license header?
Same comment for Go generator.

File: src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+			+ "# Licensed under the Apache License, Version 2.0 (the \"License\");\n"
+			+ "# you may not use this file except in compliance with the License.\n"
+			+ "# You may obtain a copy of the License at\n" + "#\n" + "#   http://www.apache.org/licenses/LICENSE-2.0\n"
+			+ "#\n" + "# Unless required by applicable law or agreed to in writing, software\n"
+			+ "# distributed under the License is distributed on an \"AS IS\" BASIS,\n"
+			+ "# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n"
+			+ "# See the License for the specific language governing permissions and\n"
+			+ "# limitations under the License.\n\n";
+
+	private final String OPTIONAL_MASSAGE = "This component is optional and therefore the API will only work if it is installed";
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java]: There's a typo (i.e. `MASSAGE`). It also affects the Go generator.

File: src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+				out.write("    # " + OPTIONAL_MASSAGE + "\n");
+			}
+		}
+
+		// Method declaration
+		out.write("    def " + createMethodName(element.getName()));
+		out.write("(");
+		
+		// Iterate through and write out method arguments
+		ArrayList<String> args = new ArrayList<String>();
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java]: The type argument could be removed from the instantiation, it's inferred by the compiler

File: src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+		if (element.getOptionalParamNames() != null) {
+			for (String param : element.getOptionalParamNames()) {
+				args.add(param.toLowerCase() + " = ''");
+			}
+		}
+		out.write(StringUtils.join(args, ", "));
+		out.write(")\n");
+
+		String method = "request";
+		if (typeOther) { method += "_other"; }
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java]: As a matter of consistency, the if statement should be indented as the other if statements (i.e. statements of if block and the closing brace should be in a new line).

File: src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+		// Return method
+		out.write("      @zap." + method + "('" + component + "/" + type + "/" + element.getName() + "/'");
+
+		// Request parameters
+		if (hasParams) {
+			out.write(", {");
+			ArrayList<String> paramArgs = new ArrayList<String>();
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java]: The same comment as before regarding the type argument.

File: src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+		}
+
+		out.write(")");
+		out.write("\n");
+		out.write("    end\n");
+		out.write("\n");
+	}
+
+	// Writes the request parameters
+	private void writeOutRequestParams(List<String> elements, ArrayList<String> args, Writer out) throws IOException {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java]: The method should be renamed, it's not doing what the name says.
The parameter `out` should be removed (as well the declared thrown exception).

File: src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java (hunk 2 above)
Comment (josuegj) [inline src/org/zaproxy/zap/extension/api/RubyAPIGenerator.java]: Red

**Comment by kingthorin** [comment]
@tjackiw anything on @thc202's feedback?

**Comment by kingthorin** [comment]
@tjackiw anything on @thc202's feedback? Should this pull be closed?

**Comment by kingthorin** [comment]
@tjackiw anything on @thc202's feedback? Should this pull be closed?

**Comment by kingthorin** [comment]
I think this one can be closed abandon? The submitter has not been in touch to follow-up and the code likely misses the mark due to 2.6.0 changes.

**Comment by thc202** [comment]
Closing, per previous comments. We can open it again if someone is interested in maintaining a Ruby client.

**Pull request closed without merging**